knora/dsplib/utils/onto_validate.py

```diff
@@ -38,15 +37,138 @@ def validate_ontology(input_file_or_json: Union[str, Dict, os.PathLike]) -> bool
+            f'$.project.ontologies[{index}].properties[?@.super[*] == hasLinkTo]'
+        ).find(data_model)
+        new = dict()
+        for match in hasLinkTo_matches:
+            prop = onto['name'] + ':' + match.value['name']
+            target = match.value['object']
+            if target != 'Resource':
+                # make the target a fully qualified name (with the ontology's name prefixed)
+                target = re.sub(r'^(:?)([^:]+)$', f'{onto["name"]}:\\2', target)
+            new[prop] = [target]
+        link_properties.update(new)
+
+    # in case the object of a property is "Resource", the link can point to every resource class
+    all_res_names: List[str] = list()
+    for index, onto in enumerate(ontos):
+        matches = jsonpath_ng.ext.parse(f'$.resources[*].name').find(onto)
+        tmp = [f'{onto["name"]}:{match.value}' for match in matches]
+        all_res_names.extend(tmp)
+    for prop, targ in link_properties.items():
+        if 'Resource' in targ:
+            link_properties[prop] = all_res_names
+
+    # make a dict that maps resource classes to their hasLinkTo-properties, and to the classes they point to
+    # example: if 'rosetta:Text' has the property 'rosetta:hasTextMedium' that points to 'rosetta:Image2D':
+    # dependencies = {'rosetta:Text': {'rosetta:hasTextMedium': ['rosetta:Image2D'], ...}}
+    dependencies: dict[str, dict[str, List[str]]] = dict()
+    for onto in ontos:
+        for resource in onto['resources']:
+            resname: str = onto['name'] + ':' + resource['name']
+            for card in resource['cardinalities']:
+                # make the cardinality a fully qualified name (with the ontology's name prefixed)
+                cardname = re.sub(r'^(:?)([^:]+)$', f'{onto["name"]}:\\2', card['propname'])
+                if cardname in link_properties:
+                    # Look out: if `targets` is created with `targets = link_properties[cardname]`, the ex-
+                    # pression `dependencies[resname][cardname] = targets` causes `dependencies[resname][cardname]`
+                    # to point to `link_properties[cardname]`. Due to that, the expression
+                    # `dependencies[resname][cardname].extend(targets)` will modify 'link_properties'!
+                    # For this reason, `targets` must be created with `targets = list(link_properties[cardname])`
+                    targets = list(link_properties[cardname])
+                    if resname not in dependencies:
+                        dependencies[resname] = dict()
+                        dependencies[resname][cardname] = targets
+                    elif cardname not in dependencies[resname]:
+                        dependencies[resname][cardname] = targets
+                    else:
+                        dependencies[resname][cardname].extend(targets)
+
+    # iteratively purge dependencies from non-circular references
+    for i in range(30):
```

Comment:
why is this 30?

Reply:
Random. It must be done a couple of times, until 'dependencies' reaches its final state and isn't changed any more by the for-loop. Perhaps it could be done more elegantly, with a while loop that checks after every pass if the pass has changed something. I tried it out quickly, but I don't see an easy solution.